python/cuda_parallel/tests/test_unique_by_key.py

```diff
@@ -0,0 +1,363 @@
+        num_items,
+    )
+
+    h_out_num_selected = d_out_num_selected.copy_to_host()
+    num_selected = h_out_num_selected[0]
+    h_out_keys = d_out_keys.copy_to_host()[:num_selected]
+    h_out_items = d_out_items.copy_to_host()[:num_selected]
+
+    expected_keys, expected_items = unique_by_key_host(
+        h_in_keys, h_in_items, compare_complex
+    )
+
+    np.testing.assert_array_equal(h_out_keys, expected_keys)
+    np.testing.assert_array_equal(h_out_items, expected_items)
+
+
+@pytest.mark.xfail(
+    reason="Creating an array of gpu_struct keys does not work currently (see https://github.com/NVIDIA/cccl/issues/3789)"
+)
+def test_unique_by_key_struct_types():
+    @gpu_struct
+    class key_pair:
+        a: np.int16
+        b: np.uint64
+
+    @gpu_struct
+    class item_pair:
+        a: np.int32
+        b: np.float32
+
+    def struct_compare_op(lhs, rhs):
+        return np.uint8((lhs.a == rhs.a) and (lhs.b == rhs.b))
+
+    num_items = 10000
+
+    a_keys = np.random.randint(0, 20, num_items, dtype=np.int16)
+    b_keys = np.random.randint(0, 20, num_items, dtype=np.uint64)
+
+    a_items = np.random.randint(0, 20, num_items, dtype=np.int32)
+    b_items = np.random.rand(num_items).astype(np.float32)
+
+    h_in_keys = np.empty(num_items, dtype=key_pair.dtype)
+    h_in_items = np.empty(num_items, dtype=item_pair.dtype)
+    h_out_num_selected = np.empty(1, np.int32)
+
+    h_in_keys["a"] = a_keys
+    h_in_keys["b"] = b_keys
+
+    h_in_items["a"] = a_items
+    h_in_items["b"] = b_items
+
+    d_in_keys = numba.cuda.to_device(h_in_keys)
+    d_in_keys = cp.asarray(d_in_keys).view(key_pair.dtype)
+    d_in_items = numba.cuda.to_device(h_in_items)
+    d_in_items = cp.asarray(d_in_items).view(item_pair.dtype)
+
+    d_out_keys = cp.empty_like(d_in_keys)
+    d_out_items = cp.empty_like(d_in_items)
+    d_out_num_selected = numba.cuda.to_device(h_out_num_selected)
+
+    unique_by_key_device(
+        d_in_keys,
+        d_in_items,
+        d_out_keys,
+        d_out_items,
+        d_out_num_selected,
+        struct_compare_op,
+        num_items,
+    )
+
+    h_out_num_selected = d_out_num_selected.copy_to_host()
+    num_selected = h_out_num_selected[0]
+    h_out_keys = d_out_keys.copy_to_host()[:num_selected]
+    h_out_items = d_out_items.copy_to_host()[:num_selected]
+
+    expected_keys, expected_items = unique_by_key_host(
+        h_in_keys, h_in_items, struct_compare_op
+    )
+
+    np.testing.assert_array_equal(h_out_keys, expected_keys)
+    np.testing.assert_array_equal(h_out_items, expected_items)
+
+
+def test_unique_by_key_with_stream(cuda_stream):
+    cp_stream = cp.cuda.ExternalStream(cuda_stream.ptr)
+    num_items = 10000
+
+    h_in_keys = random_array(num_items, np.int32, max_value=20)
+    h_in_items = random_array(num_items, np.float32)
+    h_out_keys = np.empty(num_items, dtype=np.int32)
+    h_out_items = np.empty(num_items, dtype=np.float32)
+    h_out_num_selected = np.empty(1, np.int32)
```

Comment:
Q: all examples use `np.int32` for num_selected storage. Is that what CCCL requires? If not, let's vary this type across tests.

Reply:
Varied between int32 and int64